File: src-tauri/src/domain/game.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum GameLifecycle {
    PendingSetup,
    Active,
    NeedsRepair,
    Removing,
}

impl Default for GameLifecycle {
    fn default() -> Self {
        Self::PendingSetup
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum GameHealth {
    NeedsSetup,
    Ready,
    NeedsAttention,
    Broken,
}

impl Default for GameHealth {
    fn default() -> Self {
        Self::NeedsSetup
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CloudStatus {
    Disabled,
    LocalOnly,
    Syncing,
    Synced,
    Failed,
}

impl Default for CloudStatus {
    fn default() -> Self {
        Self::Disabled
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LaunchConfig {
    pub executable_relative_path: String,
    #[serde(default)]
    pub arguments: Vec<String>,
    #[serde(default)]
    pub working_directory_relative_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Game {
    pub game_uid: String,
    pub display_name: String,
    pub managed_path: String,
    pub lifecycle: GameLifecycle,
    pub health: GameHealth,
    pub cloud_status: CloudStatus,
    pub launch: LaunchConfig,
    #[serde(default)]
    pub save_profile_id: Option<String>,
    #[serde(default)]
    pub last_played_at: Option<String>,
    #[serde(default)]
    pub latest_save_version_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GameRuntime {
    pub game_uid: String,
    pub status: GameRuntimeStatus,
    pub pid: Option<u32>,
    pub started_at: Option<String>,
    pub task_id: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum GameRuntimeStatus {
    Launching,
    Running,
    Saving,
}

impl Game {
    pub fn new_pending(
        display_name: impl Into<String>,
        managed_path: impl Into<String>,
        executable_relative_path: impl Into<String>,
    ) -> Self {
        Self {
            game_uid: Uuid::new_v4().to_string(),
            display_name: display_name.into(),
            managed_path: managed_path.into(),
            lifecycle: GameLifecycle::PendingSetup,
            health: GameHealth::NeedsSetup,
            cloud_status: CloudStatus::Disabled,
            launch: LaunchConfig {
                executable_relative_path: executable_relative_path.into(),
                ..LaunchConfig::default()
            },
            save_profile_id: None,
            last_played_at: None,
            latest_save_version_id: None,
        }
    }

    #[allow(dead_code)]
    pub fn activate(&mut self, save_profile_id: impl Into<String>) {
        self.save_profile_id = Some(save_profile_id.into());
        self.lifecycle = GameLifecycle::Active;
        self.health = GameHealth::Ready;
    }
}

#[cfg(test)]
mod tests {
    use super::{Game, GameHealth, GameLifecycle};

    #[test]
    fn new_game_requires_setup_before_activation() {
        let mut game = Game::new_pending("Test Game", "games/test", "game.exe");
        assert_eq!(game.lifecycle, GameLifecycle::PendingSetup);
        assert_eq!(game.health, GameHealth::NeedsSetup);

        game.activate("profile-1");
        assert_eq!(game.lifecycle, GameLifecycle::Active);
        assert_eq!(game.health, GameHealth::Ready);
        assert_eq!(game.save_profile_id.as_deref(), Some("profile-1"));
    }
}

File: src/domain/game.ts
export type GameLifecycle = "pending_setup" | "active" | "needs_repair" | "removing";
export type GameHealth = "needs_setup" | "ready" | "needs_attention" | "broken";
export type CloudStatus = "disabled" | "local_only" | "syncing" | "synced" | "failed";
export type GameRuntimeStatus = "launching" | "running" | "saving";

export interface LaunchConfig {
  executableRelativePath: string;
  arguments: string[];
  workingDirectoryRelativePath?: string;
}

export interface CoverCrop {
  aspectWidth: number;
  aspectHeight: number;
  outputWidth: number;
  outputHeight: number;
}

export interface CoverPosition {
  offsetXMilli: number;
  offsetYMilli: number;
  zoomMilli: number;
}

export interface GameCover {
  originalPath: string;
  displayPath: string;
  crop: CoverCrop;
  position: CoverPosition;
}

export interface Game {
  gameUid: string;
  gameKey: string;
  displayName: string;
  managedPath: string;
  lifecycle: GameLifecycle;
  health: GameHealth;
  cloudStatus: CloudStatus;
  launch: LaunchConfig;
  cover?: GameCover;
  saveProfileId?: string;
  lastPlayedAt?: string;
  latestSaveVersionId?: string;
}

export interface GameRuntime {
  gameUid: string;
  status: GameRuntimeStatus;
  pid?: number;
  startedAt?: string;
  taskId?: string;
}

export interface GameBodyVersion {
  versionId: string;
  gameUid: string;
  createdAt: string;
  archivePath: string;
  fileCount: number;
  totalBytes: number;
  packagePath?: string;
  packageSize?: number;
  sha256?: string;
  excludedItems: string[];
  uploadStatus?: string;
  remotePath?: string;
  remoteFsId?: number;
  remoteSize?: number;
}

export interface LaunchPrecheck {
  gameUid: string;
  canLaunch: boolean;
  executableExists: boolean;
  saveProfileReady: boolean;
  validScopeCount: number;
  issues: string[];
}

export interface SaveFileEntry {
  rootType: SaveRootType;
  rootPath?: string;
  relativePath: string;
  objectHash?: string;
  size: number;
  deleted: boolean;
}

export interface SaveVersion {
  versionId: string;
  gameUid: string;
  createdAt: string;
  files: SaveFileEntry[];
  totalBytes: number;
}

export interface SaveProfile {
  profileId: string;
  gameUid: string;
  executableHash: string;
  scopes: SaveScope[];
  detectionEvidence: string[];
  confidence: number;
  enabled: boolean;
  keepVersions?: number;
  createdAt: string;
  updatedAt: string;
}

export type SaveRootType = "managed_game" | "app_data" | "local_app_data" | "local_low" | "documents" | "saved_games" | "user_profile" | "custom";
export type UnknownFilePolicy = "protect" | "ignore";

export interface SaveScope {
  rootType: SaveRootType;
  rootPath: string;
  confirmedFiles: string[];
  includeDirectories: string[];
  excludeExact: string[];
  excludePatterns: string[];
  excludeDirectories: string[];
  unknownFilePolicy: UnknownFilePolicy;
  maxFileBytes?: number;
}

export interface SaveScopeDraft {
  scope: SaveScope;
  changedFiles: string[];
  confidence: number;
}

export interface SaveLearningSession {
  sessionId: string;
  gameUid: string;
  rootPid: number;
  startedAt: string;
  status: "capturing" | "finished" | "cancelled";
}

export interface SaveLearningResult {
  sessionId: string;
  changedFiles: string[];
  scopeDrafts: SaveScopeDraft[];
  confidence: number;
  notes: string[];
  eventCaptureMode: "etw" | "snapshot" | string;
  transactionSummary?: SaveTransactionSummary;
}

export interface SaveTransactionSummary {
  status: "completed" | "candidate" | "insufficient_evidence" | string;
  confidence: number;
  transactionCount: number;
  affectedFiles: string[];
  affectedDirectories: string[];
  startedAt?: string;
  endedAt?: string;
  operationCount: number;
  notes: string[];
}

export interface CloudSaveManifestVersion {
  versionId: string;
  createdAt: string;
  packagePath: string;
  packageFsId: number;
  packageSize: number;
  packageSha256?: string;
  fileCount: number;
  totalBytes: number;
  deviceName?: string;
}

export interface CloudSaveSyncStatusView {
  autoSyncSave: boolean;
  localVersionCount: number;
  cloudVersionCount: number;
  latestLocalCreatedAt?: string;
  latestCloudCreatedAt?: string;
  latestCloudVersionId?: string;
  syncState: "synced" | "local_ahead" | "cloud_ahead" | "no_cloud_saves" | "offline" | string;
  warnings: string[];
}

export function gameStatusLabel(game: Game): string {
  if (game.lifecycle === "pending_setup") return "需要设置";
  if (game.lifecycle === "needs_repair" || game.health === "broken" || game.health === "needs_attention") {
    return "需要处理";
  }
  return "可启动";
}
